url882f91ec97_WenboCao_Microsoft_Drone_tgz-pJ8-org_ejml_alg_dense_decomposition_qr_StabilityQRDecompositionJ8/src/org/ejml/alg/block/TestBlockInnerTriangularSolver.java

```diff
@@ -0,0 +1,185 @@
+    }
+
+    @Test
+    public void testInvertLower_one() {
+        DenseMatrix64F A = RandomMatrices.createUpperTriangle(5,0,-1,1,rand);
+        CommonOps.transpose(A);
+
+        DenseMatrix64F A_inv = A.copy();
+
+        BlockInnerTriangularSolver.invertLower(A_inv.data,5,0);
+
+        DenseMatrix64F S = new DenseMatrix64F(5,5);
+        CommonOps.mult(A,A_inv,S);
+
+        assertTrue(GenericMatrixOps.isIdentity(S,1e-8));
+    }
+
+    /**
+     * Test all inner block solvers using reflections to look up the functions
+     */
+    @Test
+    public void testSolveArray() {
+        Method methods[] = BlockInnerTriangularSolver.class.getMethods();
+
+        int numFound = 0;
+        for( Method m : methods) {
+            String name = m.getName();
+
+            if( !name.contains("solve") || name.compareTo("solve") == 0 || name.compareTo("solveBlock") == 0 )
+                continue;
+
+//            System.out.println("name = "+name);
+
+            boolean solveL = name.contains("L");
+            boolean transT;
+            boolean transB = name.contains("TransB");
+
+            if( solveL )
+                transT = name.contains("TransL");
+            else
+                transT = name.contains("TransU");
+
+            check_solve_array(m,solveL,transT,transB);
+
+            numFound++;
+        }
+
+        // make sure all the functions were in fact tested
+        assertEquals(5,numFound);
+    }
+
+    /**
+     * Checks to see if solve functions that use arrays as input work correctly.
+     */
+    private void check_solve_array(Method m,
+                                   boolean solveL, boolean transT, boolean transB) {
+        int offsetL = 2;
+        int offsetB = 3;
+
+        DenseMatrix64F L = createRandomLowerTriangular(3);
+
+        if( !solveL ) {
+            CommonOps.transpose(L);
+        }
+
+        if( transT ) {
+            CommonOps.transpose(L);
+        }
+
+        DenseMatrix64F L_inv = L.copy();
+        UnrolledInverseFromMinor.inv(L_inv,L_inv);
+
+        DenseMatrix64F B = RandomMatrices.createRandom(3,4,rand);
+        DenseMatrix64F expected = RandomMatrices.createRandom(3,4,rand);
+        DenseMatrix64F found = B.copy();
+
+        // compute the expected solution
+        CommonOps.mult(L_inv,B,expected);
+
+        if( transT ) {
+            CommonOps.transpose(L);
+        }
+
+        if( transB ) {
+            CommonOps.transpose(found);
+            CommonOps.transpose(expected);
+        }
+
+        // create arrays that are offset from the original
+        // use two different offsets to make sure it doesn't confuse them internally
+        double dataL[] = offsetArray(L.data,offsetL);
+        double dataB[] = offsetArray(found.data,offsetB);
+
+        try {
+            m.invoke(null,dataL,dataB,3,4,3,offsetL,offsetB);
+        } catch (IllegalAccessException e) {
+            fail("invoke failed");
+        } catch (InvocationTargetException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)